universe/ValueRefs.cpp

```diff
@@ -285,20 +286,22 @@ std::string ValueRefBase::InvariancePattern() const {
         .append(ConstantExpr()                  ? "C" : "c");
 }
 
-MeterType NameToMeter(const std::string& name) {
-    MeterType retval = MeterType::INVALID_METER_TYPE;
-    auto it = GetMeterNameMap().find(name);
-    if (it != GetMeterNameMap().end())
-        retval = it->second;
-    return retval;
+constexpr MeterType NameToMeter(const std::string_view name) {
+    for (int i = 0; i < static_cast<int>(NAME_BY_METER.size()); i++) {
+        if (NAME_BY_METER[i] == name)
+            return static_cast<MeterType>(i - 1);
+    }
+
+    return MeterType::INVALID_METER_TYPE;
 }
 
-const std::string& MeterToName(MeterType meter) {
-    for (auto& [name, type] : GetMeterNameMap()) {
-        if (type == meter)
-            return name; // CodeQL reports "Returning stack-allocated memory", but the name structured binding should be a reference as far as I can tell
-    }
-    return EMPTY_STRING;
+static_assert(NameToMeter("not a meter") == MeterType::INVALID_METER_TYPE, "Name to Meter conversion failed for invalid meter type!");
+static_assert(NameToMeter("Population") == MeterType::METER_POPULATION, "Name to Meter conversion failed for 'Population' meter!");
+static_assert(NameToMeter("Speed") == MeterType::METER_SPEED, "Name to Meter conversion failed for 'Speed' meter!");
```

Comment:
Seem strange that it fails.

Reply:
This is mostly future proofing if more meters are added and/or the enum order changes I think.

Reply:
Such things should be in tests and current implementation is error prone in future code edit.

Reply:
There is nothing wrong with using `static_assert` for testing when appropriate. If a code edit breaks it, then the test will fail.

Reply:
In [TestEnumParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestEnumParser.cpp) there are tests for each Enum like:
```
CHECK_ENUM_AND_RESULT("TargetPopulation", MeterType::METER_TARGET_POPULATION, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetPopulation", MeterType::METER_TARGET_POPULATION, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetIndustry", MeterType::METER_TARGET_INDUSTRY, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetIndustry", MeterType::METER_TARGET_INDUSTRY, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetResearch", MeterType::METER_TARGET_RESEARCH, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetResearch", MeterType::METER_TARGET_RESEARCH, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetInfluence", MeterType::METER_TARGET_INFLUENCE, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetInfluence", MeterType::METER_TARGET_INFLUENCE, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetConstruction", MeterType::METER_TARGET_CONSTRUCTION, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetConstruction", MeterType::METER_TARGET_CONSTRUCTION, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxFuel", MeterType::METER_MAX_FUEL, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxFuel", MeterType::METER_MAX_FUEL, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxShield", MeterType::METER_MAX_SHIELD, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxShield", MeterType::METER_MAX_SHIELD, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxStructure", MeterType::METER_MAX_STRUCTURE, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxStructure", MeterType::METER_MAX_STRUCTURE, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxDefense", MeterType::METER_MAX_DEFENSE, MeterType, parse::non_ship_part_meter_enum_grammar);

```

In [TestValueRefDoubleParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestValueRefDoubleParser.cpp) there is example of parsing math example and result of AbstractSyntaxTree which tell compiler how to know what it means:
```
/*
Term:
  -1.5+2.9/-7.4

Expected AST:
  PLUS
    NEGATE
      1.5
    DIVIDE
      2.9
      NEGATE
        7.4
*/
```

Such tests should also be doable for something **as seen in [TestEnumParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestEnumParser.cpp) comments**(and when that happens, position in array should be problem, but tests of enum are in place and missing should be added in future(**but adding it in wrong order cause test to fail as it should**))´:
```
// XXX: METER_SIZE not handled, still used?
// XXX: HAPPINESS meters not handled, still used?
// XXX: REBEL_TROOPS meters not handled, still used?
// XXX: No enum number value to validate enum coverage.
//      Maybe the Meter enum should be split.
```

Therefore AbstractSyntaxTree for something like([Python part tech](https://github.com/freeorion/freeorion/blob/master/default/scripting/techs/ship_parts/fuel/fuel.py)):
```
def PART_UPGRADE_MAXFUEL_EFFECTS(effect_name: str, value: float):
    return EffectsGroup(
        scope=Ship & OwnedBy(empire=Source.Owner) & DesignHasPart(name="FU_BASIC_TANK"),
        accountinglabel=effect_name,
        effects=[
            SetMaxFuel(
                value=Value
                + NamedReal(name=effect_name + "_MULT", value=value)
                * PartsInShipDesign(name="FU_BASIC_TANK", design=Target.DesignID)
            )
        ],
    )

```

SetMaxFuel:
 previous_value + some_effect * ship_parts_capacity


When wanting in [ValueRefs.cpp](https://github.com/freeorion/freeorion/blob/master/universe/ValueRefs.cpp) convert and evaluate some string_view to do some (meter) type and viceversa, there are also **invariants** (=something that is always true)(or candidate)

**For a function to be fit for use in constant expressions, it must be explicitly declared constexpr**. From [What's the difference between constexpr and const?](https://stackoverflow.com/questions/14116003/whats-the-difference-between-constexpr-and-const)